### DISCUSSION TITLE
Newtonsoft bindingRedirect via manifest configuration component

### DISCUSSION THREAD

**Discussion by DNNMonster** [dnnsoftware/Dnn.Platform#4295] [Q&A] [answered]
Newtonsoft(12.0.3) is a couple major versions ahead of DNN(10.0.3) now.  Can someone share the proper xpath & syntax required by the manifest configuration component to create the following during a module install:

```
<dependentAssembly xmlns="urn:schemas-microsoft-com:asm.v1">
   <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" />
   <bindingRedirect oldVersion="0.0.0.0-10.0.3.32767" newVersion="10.0.0.0" />
   <bindingRedirect oldVersion="10.0.4.0-32767.32767.32767.32767" newVersion="12.0.0.0" />
   <codeBase version="12.0.0.0" href="bin\NewtonSoft.Json\V12\Newtonsoft.Json.dll" />
   <codeBase version="10.0.0.0" href="bin\Newtonsoft.Json.dll" />
</dependentAssembly>
```

If it's not possible to generate that syntax via the manifest configuration component during a module installation that would be good to hear as well.

I've tried multiple permutations but they all leave only the binding to the v12.0.3 as below which crashes an install.  

```
<dependentAssembly xmlns="urn:schemas-microsoft-com:asm.v1">
        <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" />
        <bindingRedirect oldVersion="0.0.0.0-32767.32767.32767.32767" newVersion="12.0.0.0" />
      </dependentAssembly>
```

The module project I'm looking at references a 3rd party library which depends on Newtonsoft v12.0.3.

The code below ***does NOT*** work but it may help expose the error in my approach.  

```
<component type="Config">
          <config>
            <configFile>web.config</configFile>
            <install>
              <configuration>
                <nodes>
                  <node path="/configuration/runtime/assemblyBinding" 
                        action="update" 
                        targetpath="/configuration/runtime/assemblyBinding/dependentAssembly/assemblyIdentity [@name='Newtonsoft.json']" 
                        collision="save">

                    <bindingRedirect oldVersion="1.0.0.0-10.0.3.32767" newVersion="10.0.0.0" />
                    <bindingRedirect oldVersion="10.0.4.0-32767.32767.32767.32767" newVersion="12.0.0.0" />
                    <codeBase version="12.0.0.0" href="bin\NewtonSoft.Json\V12\Newtonsoft.Json.dll" />
                    <codeBase version="10.0.0.0" href="bin\Newtonsoft.Json.dll" />
                  </node>

                </nodes>
              </configuration>
            </install>
            <uninstall>
              
                <!-- TODO: Consider removing config entries & Uninstalling the 
                            bin/Newtonsoft.json/v12 directory -->
                
            </uninstall>
          </config>
        </component>
```

**Comment by mitchelsellers** [comment]
What is the error that you get with the wide-sweeping rule?

(NOTE: This is NOT a supported, nor-suggested, process to override and could very well have major impacts to DNN functionality)

**Comment by DNNMonster** [comment]
Could not load file or assembly 'Newtonsoft.Json' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

[FileLoadException: Could not load file or assembly 'Newtonsoft.Json' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]

[FileLoadException: Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]
   System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks) +0
   System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks) +232
   System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection) +113
   System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection) +23
   System.Reflection.Assembly.Load(String assemblyString) +35
   System.Web.Configuration.CompilationSection.LoadAssemblyHelper(String assemblyName, Boolean starDirective) +49

[ConfigurationErrorsException: Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]
   System.Web.Configuration.CompilationSection.LoadAssemblyHelper(String assemblyName, Boolean starDirective) +762
   System.Web.Configuration.CompilationSection.LoadAllAssembliesFromAppDomainBinDirectory() +259
   System.Web.Configuration.CompilationSection.LoadAssembly(AssemblyInfo ai) +167
   System.Web.Compilation.BuildManager.GetReferencedAssemblies(CompilationSection compConfig) +238
   System.Web.Compilation.BuildProvidersCompiler..ctor(VirtualPath configPath, Boolean supportLocalization, String outputAssemblyName) +62
   System.Web.Compilation.CodeDirectoryCompiler.GetCodeDirectoryAssembly(VirtualPath virtualDir, CodeDirectoryType dirType, String assemblyName, StringSet excludedSubdirectories, Boolean isDirectoryAllowed) +658
   System.Web.Compilation.BuildManager.CompileCodeDirectory(VirtualPath virtualDir, CodeDirectoryType dirType, String assemblyName, StringSet excludedSubdirectories) +176
   System.Web.Compilation.BuildManager.CompileResourcesDirectory() +50
   System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled() +297

[HttpException (0x80004005): Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]
   System.Web.Compilation.BuildManager.ReportTopLevelCompilationException() +76
   System.Web.Compilation.BuildManager.EnsureTopLevelFilesCompiled() +631
   System.Web.Compilation.BuildManager.CallAppInitializeMethod() +38
   System.Web.Hosting.HostingEnvironment.Initialize(ApplicationManager appManager, IApplicationHost appHost, IConfigMapPathFactory configMapPathFactory, HostingEnvironmentParameters hostingParameters, PolicyLevel policyLevel, Exception appDomainCreationException) +758

[HttpException (0x80004005): Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]
   System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +552
   System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +122
   System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +737

**Comment by mitchelsellers** [comment]
Is it possible that you are using a 32 bit version of the dll?

**Reply by DNNMonster** [reply]
Name                  : Newtonsoft.Json
Version               : 12.0.0.0
CultureInfo           :
CultureName           :
CodeBase              : file:///C:/inetpub/wwwroot/dnndev3/bin/Newtonsoft.Json/v12/Newtonsoft.Json.dll
EscapedCodeBase       : file:///C:/inetpub/wwwroot/dnndev3/bin/Newtonsoft.Json/v12/Newtonsoft.Json.dll
ProcessorArchitecture : MSIL
ContentType           : Default
Flags                 : PublicKey
HashAlgorithm         : SHA1
VersionCompatibility  : SameMachine
KeyPair               :
FullName              : Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed

**Comment by DNNMonster** [comment]
What's the proper way to run side by side versions of a library without using bindingRedirects?

**Comment by mitchelsellers** [comment]
We do not have any suggested patterns, as they all have various negative drawbacks.  The two primary methods that I've seen used.

* Custom Probing Paths - To load from a sub-folder of the bin, or other location.
* Reflection Loading in a Secondary App Domain - This has some huge memory considerations if you don't know what you are doing

I wish I could be more help here, but this is a very advanced, and sticky situation.  I'll see if I can get a few of those that do this to jump in and give you some pointers.

**Comment by bdukes** [comment]
See [`BindingRedirect.config`](https://github.com/dnnsoftware/Dnn.Platform/blob/9ff814db2728da79ab38a092225cb8e5182038e0/DNN%20Platform/Website/Install/Config/BindingRedirect.config), which is what the platform uses to add binding redirects during module installs.  You'll want to add the namespaces, and adjust your `targetpath` attribute to point to the `dependentAssembly` element instead of the `assemblyIdentity` element.

That said, I think that DNN would overwrite whatever change you make when it upgrades Newtonsoft.

**Comment by DNNMonster** [comment]
Thanks for the responses.  I believe DNN is bumping the Newtonsoft version around DNN 10..   That's not too far off. 

 I understand the config overwrites during an upgrade will be an issue until it overwrites to v12+ at which point I am guessing all will be fine and the bindingRedirects could be removed.  I think the best I can do here is have admins do a module re-install/repair after a platform upgrade in the meantime to ensure a web config that will work for the module.   That process is familiar to all DNN admins and the module I'm looking at would affect a specific workflow which is not cool...  but it would not mean a system crash if an admin forgot to immediately do a post platform upgrade module re-install.

Here is code that I am testing that appears to be working as desired.  I think with a few minor tweaks it will work for any other side by side versioning issues as well.   Let me know if you see some issue here other than the upgrade overwrite.   
```

<component type="Config">
          <config>
            <configFile>web.config</configFile>
            <install>
              <configuration>
                <nodes>
                  <node path="/configuration/runtime/ab:assemblyBinding" 
                        action="update" 
                        targetpath="/configuration/runtime/ab:assemblyBinding/ab:dependentAssembly[ab:assemblyIdentity/@name='Newtonsoft.Json']" 
                        nameSpace="urn:schemas-microsoft-com:asm.v1" 
                        nameSpacePrefix="ab" 
                        collision="save">
                    <dependentAssembly xmlns="urn:schemas-microsoft-com:asm.v1">
                      <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" />
                      <bindingRedirect oldVersion="1.0.0.0-10.0.3.32767" newVersion="10.0.0.0" />
                      <bindingRedirect oldVersion="10.0.4.0-32767.32767.32767.32767" newVersion="12.0.0.0" />
                      <codeBase version="12.0.0.0" href="bin\NewtonSoft.Json\V12\Newtonsoft.Json.dll" />
                      <codeBase version="10.0.0.0" href="bin\Newtonsoft.Json.dll" />
                    </dependentAssembly>
                  </node>
                </nodes>
              </configuration>
            </install>
            <uninstall>
              
                <!-- TODO: Consider removing config entries & Uninstalling the 
                            bin/Newtonsoft.json/v12 directory -->
                
            </uninstall>
          </config>
        </component>
```

**Comment by bdukes** [comment, marked as answer]
You'll need to set `xmlns="urn:schemas-microsoft-com:asm.v1"` on the elements in your `<node>` element (or target the `dependentAssembly` element instead of the `bindingRedirect`, [like the framework does](https://github.com/dnnsoftware/Dnn.Platform/blob/9ff814db2728da79ab38a092225cb8e5182038e0/DNN%20Platform/Website/Install/Config/BindingRedirect.config), but you still need to specify it somewhere)

**Reply by DNNMonster** [reply]
I edited the post to reflect your suggestions and not leave bogus code hanging around.  It looks right now and I am not seeing any side effects.  Thanks!
